src/core/lib/gprpp/sync.h

```diff
@@ -0,0 +1,79 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* Inline implementation of synchronization primitives, preferred over the C
+   public API for C++ code. */
+
+#ifndef GRPC_CORE_LIB_GPRPP_SYNC
+#define GRPC_CORE_LIB_GPRPP_SYNC
+
+#include <grpc/support/port_platform.h>
+
+#include <grpc/support/atm.h>
+#include <grpc/support/log.h>
+#include <grpc/support/sync.h>
+
+#include <assert.h>
+
+namespace grpc_core {
+
+inline void RefInit(gpr_refcount* r, int n) {
```

Comment:
This is a brilliant move... but can I suggest that since this is all for deep-core use, is there any reason that we need to use the API structs of gpr_refcount? Or can this instead be a full C++ class with normal methods, etc? The free functions feel a little unusual.

Reply:
thank you @vjpai! Sure, I can implement a class for this.

I added the free functions to share the same implementation with the C `gpr_(ref|unref)*` helper. In other words, if I add a C++ class without the free functions, the implementation wouldn't be shared with the C API.

So, to create a C++ class, I can think of two options:
1) Introduce a new C++ class and use it everywhere in C++ code. Keep the C API for `gpr_refcount` untouched.
2) Get rid of `gpr_refcount` from the public API (assuming there is no backwards compatibility concerns) and use the new C++ class.

Which one would you prefer?

Reply:
I was just looking over this. I am ok with option 2 because gpr_refcount is not used by any of the wrapped languages (only by core and C++).  I would want @nicolasnoble 's blessing to do this, but I vote to kill it (which will require a gRFC and a major version bump for core).

Reply:
Same thing with stats_counter ; it is only used in core and C++, so let's kill them all.

Reply:
Thank you @vjpai! Sounds great. As per our offline discussion, I'm going to do this in two steps:
Step 1) we will add an idiomatic C++ implementations in the first PR and convert all uses of gpr_refcount and statscounter to the new implementation.
Step 2) we will write a proposal to remove it from the public API and will proceed with killing the interface with a follow up PR.